### PR TITLE
feat(input-time): calcite-input-time component

### PR DESCRIPTION
**Pull request by eriklharper**
**Related Issue:** #2709 

## Summary

Introduces the new `calcite-input-time` component for simplified time entry.

### CODE REVIEW

File: src/components/input-time/input-time.tsx [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+    }
+    if (composedEventPath.includes(this.minuteEl)) {
+      return;
+    }
+    if (composedEventPath.includes(this.secondEl)) {
+      return;
+    }
+    if (composedEventPath.includes(this.meridiemEl)) {
+      return;
+    }
```
Comment (benelan) [inline src/components/input-time/input-time.tsx:281]: Can you use one if statement for all of these, it seems overly verbose. Maybe just a bunch of `||`'s, or throw the props in any array and use `some()`.

```js
if ([this.hourEl, this.minuteEl, this.secondEl, this.meridiemEl].some((el) =>
    composedEventPath.includes(el))) return;
```

`||` is better performance wise though

File: src/components/input-time/input-time.tsx [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+  //--------------------------------------------------------------------------
+
+  /**
+   * Sets focus on the component.
+   *
+   * @param target
+   */
+  @Method()
+  async setFocus(target?: TimePart): Promise<void> {
+    this[`${target || "hour"}El`]?.focus();
```
Comment (benelan) [inline src/components/input-time/input-time.tsx:357]: I forget the name, but you need to await a util from `loadable.ts` before calling `focus()`. See any other components for an example.

File: src/components/input-time/input-time.tsx [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+      }
+    }
+    this.localizedMeridiem = this.value
+      ? localizeTimeStringToParts({ value: this.value, locale })?.localizedMeridiem || null
+      : localizeTimePart({ value: this.meridiem, part: "meridiem", locale });
+  };
+
+  private timePartFocusHandler = (event: FocusEvent): void => {
+    this.activeEl = event.currentTarget as HTMLSpanElement;
+  };
```
Comment (benelan) [inline src/components/input-time/input-time.tsx:723]: Oh man, there's lot of if/else and case/switch statements in all the methods and handlers above lol 😵
It would be nice to refactor this at some point to make it more readable, maybe create some functional utils or something.

File: src/components/input-time/input-time.tsx [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+          )}
+          {this.localizedSecondSuffix && (
+            <span class={CSS.delimiter}>{this.localizedSecondSuffix}</span>
+          )}
+          {showMeridiem && (
+            <span
+              aria-label={this.intlMeridiem}
+              aria-valuemax="2"
+              aria-valuemin="1"
+              aria-valuenow={(this.meridiem === "PM" && "2") || "1"}
```
Comment (benelan) [inline src/components/input-time/input-time.tsx:850]: We probably don't want to hardcode "PM" right, since we do translations?

**Review by benelan — CHANGES_REQUESTED** [review]
Here are some comments from a quick first pass. I'll do some demo testing and a more in depth review Friday morning after you have time to work through the feedback.

File: src/components/input-time/input-time.tsx [hunk 5]
```diff
@@ -0,0 +1,872 @@
...
+            [CSS.readonly]: this.readonly,
+            [CSS[`scale-${this.scale}`]]: true,
+            [CSS.showMeridiem]: showMeridiem,
+            [CSS.showSecond]: this.showSecond
+          }}
+          dir="ltr"
+        >
+          <calcite-icon class={CSS.clockIcon} flipRtl={this.iconFlipRtl} icon="clock" scale="s" />
+          <span
+            aria-label={this.intlHour}
```
Comment (anveshmekala) [inline src/components/input-time/input-time.tsx:782]: This needs to use `t9nComponent` interface . `intl-*` pattern is deprecated.

Comment (anveshmekala) [reply src/components/input-time/input-time.tsx:782]: applies to other instances as well.

File: src/components/input-time/input-time.tsx [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+   * @default "AM/PM"
+   */
+  @Prop() intlMeridiem = TEXT.meridiem;
+
+  /**
+   * aria-label for the minute input
+   *
+   * @default "Minute"
+   */
+  @Prop() intlMinute = TEXT.minute;
```
Comment (anveshmekala) [inline src/components/input-time/input-time.tsx:104]: any label's that needs to be translated should use `messages` & `messageOverrides` properties. Please ref component with `t9nComponent` interface.

File: package-lock.json [hunk 7]
```diff
@@ -21,7 +21,7 @@
       "devDependencies": {
```
Comment (alisonailea) [inline package-lock.json:21]: I think this file should be reverted and the lock file changes done in a separate PR or a publish.

File: src/components/input-time/input-time.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+  focusable,
+  formAssociated,
+  labelable,
+  reflects,
+  renders
+} from "../../tests/commonTests";
+import { CSS } from "./resources";
+
+describe("calcite-input-time", () => {
+  it("renders", async () => renders("calcite-input-time", { display: "inline-block" }));
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:16]: suggestion: group all tests run by commonTests resources under a `describe("common tests", () => {...`

File: src/components/input-time/input-time.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+  it("is labelable", async () => labelable("calcite-input-time"));
+
+  it("should focus the hour input when setFocus is called", async () =>
+    focusable(`calcite-input-time`, {
+      shadowFocusTargetSelector: "span.hour"
+    }));
+
+  it("can be disabled", () => disabled("calcite-input-time"));
+
+  it("when set to readOnly, element still focusable but won't display the controls or allow for changing the value", async () => {
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:49]: can you rephrase this using the `it("should..."` convention?

File: src/components/input-time/input-time.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+    const page = await newE2EPage();
+    await page.setContent(`<calcite-input-time id="canReadOnly" readonly></calcite-input-time>`);
+
+    const component = await page.find("calcite-input-time");
+
+    await component.callMethod("setFocus");
+    await page.waitForChanges();
+
+    expect(await page.evaluate(() => document.activeElement.id)).toBe("canReadOnly");
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:58]: this `expect` is unnecessary and it's validating anything from the component so it should be removed

File: src/components/input-time/input-time.e2e.ts [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+    accessible(`
+    <calcite-label>
+      Input Time
+      <calcite-input-time name="test"></calcite-input-time>
+    </calcite-label>
+  `));
+
+  it("has defaults", async () =>
+    defaults("calcite-input-time", [
+      { propertyName: "scale", defaultValue: "m" },
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:28]: does it make sense to standardize around storing defaults in `./resources.ts` so they can be referenced in the component and tests?

File: src/components/input-time/input-time.e2e.ts (hunk 10 above)
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:58]: can you move this expect to the bottom of the test. I'd like to group the expects as much as possible for easier readability.

File: src/components/input-time/input-time.e2e.ts [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    await page.waitForChanges();
+
+    await component.type("12");
+    await page.keyboard.press("Tab");
+    await component.type("30");
+    await page.waitForChanges();
+
+    expect(await component.getProperty("value")).toBeNull();
+  });
+
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:70]: group these lang specific tests under a describe. Something like... `describe("lang specific UI value changes", () => { ...`

File: src/components/input-time/input-time.e2e.ts [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+    expect(inputEvent).toHaveReceivedEventTimes(2);
+    expect(changeEvent).toHaveReceivedEventTimes(1);
+
+    await page.keyboard.down("Shift");
+    await page.keyboard.press("Tab");
+    await page.keyboard.up("Shift");
+    await page.keyboard.press("Tab");
+    await page.waitForChanges();
+
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:288]: you can remove all the `expect`s before this one to improve the speed of the tests without losing test integrity. If you want to better describe what is happening, add some inline comments.

File: src/components/input-time/input-time.e2e.ts [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+    await page.keyboard.press("Enter");
+    await page.waitForChanges();
+
+    expect(inputEvent).toHaveReceivedEventTimes(1);
+    expect(changeEvent).toHaveReceivedEventTimes(1);
+
+    await page.keyboard.press("Enter");
+    await page.waitForChanges();
+
+    expect(inputEvent).toHaveReceivedEventTimes(1);
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:323]: same as test above. Let's improve test performance by removing unnecessary `expect`. Add some inline comments to describe when the input event is called and why you can press the enter key multiple times and the change event only calls once.

File: src/components/input-time/input-time.e2e.ts [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+  it("sets initial value to null when it is not a valid time value", async () => {
+    const page = await newE2EPage({
+      html: `<calcite-input-time value="invalid"></calcite-input-time>`
+    });
+    const inputTime = await page.find("calcite-input-time");
+
+    expect(await inputTime.getProperty("value")).toBeNull();
+  });
+
+  it("is form-associated", () => formAssociated("calcite-input-time", { testValue: "03:23", submitsOnEnter: true }));
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:359]: move this test up with the other common tests

File: src/components/input-time/input-time.e2e.ts [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+    const inputTime = await page.find("calcite-input-time");
+
+    await page.evaluate(() => {
+      const inputTime = document.querySelector("calcite-input-time");
+      inputTime.addEventListener("calciteInputTimeInput", (event) => {
+        event.preventDefault();
+      });
+    });
+
+    expect(await inputTime.getProperty("value")).toBe("14:59");
```
Comment (alisonailea) [inline src/components/input-time/input-time.e2e.ts:340]: remove this unnecessary `expect`

File: src/utils/time.ts [hunk 17]
```diff
@@ -66,12 +66,42 @@ export function formatTimeString(value: string): string {
...
+  const minuteIndex = parts.indexOf(parts.find(({ type }): boolean => type === "minute"));
+  const secondIndex = parts.indexOf(parts.find(({ type }): boolean => type === "second"));
+  const minuteSuffix = parts[minuteIndex + 1];
+  return minuteSuffix && minuteSuffix.type === "literal" && secondIndex - minuteIndex === 2
+    ? minuteSuffix.value?.trim() || null
+    : null;
+}
+
+export function getLocaleSecondSuffix(locale: string): string {
+  const formatter = createLocaleDateTimeFormatter(locale);
```
[the remark this answers is not in the document]
Comment (alisonailea) [reply src/utils/time.ts:98]: this should probably be done once in connectedCallback or componentWillLoad. if the user is changing the locale, the whole page is going to be reloading anyway.

File: src/components/input-time/input-time.tsx [hunk 18]
```diff
@@ -0,0 +1,872 @@
...
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Properties
+  //
+  // --------------------------------------------------------------------------
+
+  private activeEl: HTMLSpanElement;
+
+  defaultValue: InputTime["value"];
```
Comment (alisonailea) [inline src/components/input-time/input-time.tsx:174]: if these are private, please add `private` in front of them

File: src/components/input-time/input-time.tsx [hunk 19]
```diff
@@ -0,0 +1,872 @@
...
+  private secondEl: HTMLSpanElement;
+
+  private userChangedValue = false;
+
+  // --------------------------------------------------------------------------
+  //
+  //  State
+  //
+  // --------------------------------------------------------------------------
+
```
Comment (alisonailea) [inline src/components/input-time/input-time.tsx:201]: Is there a way to simplify these State props? It seems like we only need the localized state or regular state, not both as updating one could drive the other.

File: src/components/input-time/input-time.tsx [hunk 20]
```diff
@@ -0,0 +1,872 @@
...
+   */
+  @Event() calciteInternalInputTimeFocus: EventEmitter<void>;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Event Listeners
+  //
+  //--------------------------------------------------------------------------
+
+  @Listen("blur")
```
Comment (alisonailea) [inline src/components/input-time/input-time.tsx:260]: we need to create some documentation alignment on how we are handling event listeners. I prefer this syntax but other devs are saying these should be setup as on event listeners in the component render itself. @benelan thoughts?

File: src/components/input-time/input-time.tsx (hunk 1 above)
Comment (alisonailea) [reply src/components/input-time/input-time.tsx:281]: the performance difference is very slight however so in this case I would defer to which seems more readable.

File: src/components/input-time/input-time.tsx [hunk 21]
```diff
@@ -0,0 +1,872 @@
...
+          break;
+        case " ":
+        case "Spacebar":
+          event.preventDefault();
+          break;
+      }
+    }
+  };
+
+  onLabelClick(): void {
```
Comment (alisonailea) [inline src/components/input-time/input-time.tsx:566]: private?

File: src/components/input-time/input-time.tsx (hunk 3 above)
Comment (alisonailea) [reply src/components/input-time/input-time.tsx:723]: agreed. I'd like to start pulling as many private functions out into their own single testable files in a ./[component]/functions/[fileName]` pattern. It will let us better test our components and maybe recognize more functional patterns we already have.

**Review by alisonailea — CHANGES_REQUESTED** [review]
Left a few comments. Major priorities are 
1. organizing e2e tests
2. think about how much of the input-time component can be written as functional utils/component(s).

File: src/utils/time.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (benelan) [reply src/utils/time.ts:98]: The whole page won't necessarily be reloaded if the user changes the `lang` property on the specific component.

File: src/components/input-time/input-time.tsx (hunk 20 above)
Comment (benelan) [reply src/components/input-time/input-time.tsx:260]: Yeah I don't feel strongly in either direction but I agree on the need for consistency. I'll run it by Franco when he returns.

File: src/components/input-time/resources.ts [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+  second: "second",
+  showMeridiem: "show-meridiem",
+  showSecond: "show-second"
+};
+
+export const TEXT = {
+  hour: "Hour",
+  meridiem: "AM/PM",
+  minute: "Minute",
+  second: "Second"
```
[the remark this answers is not in the document]
Comment (anveshmekala) [reply src/components/input-time/resources.ts:24]: yup we have these translated for `calcite-time-picker`. currently, we don't have shared assets set up yet, so we have to copy the message bundles from time-picker component and generate the type using `util:generate-t9n-types` command. Please make sure that the component path is added in `manifest` file.

File: src/components/input-time/input-time.e2e.ts (hunk 11 above)
Comment (benelan) [reply src/components/input-time/input-time.e2e.ts:28]: Yeah I like that idea

**Comment by eriklharper** [comment]
Closing in favor of https://github.com/Esri/calcite-design-system/issues/8624
